### DISCUSSION TITLE
[Question] New static column in IStorage.

### DISCUSSION THREAD

**Discussion by crawe1** [YetiForceCompany/YetiForceCRM#14056] [General]
Hi,
I want to add new static column in table. Someone know how can i add static column in storage module ?

![image](https://user-images.githubusercontent.com/61950059/104828272-d39ba200-5867-11eb-8ad5-6ff0c61095e4.png)

for example i want to duplicate only this column but this is realated istorage qty but how this Display?

**Comment by crawe1** [comment]
ok its simple for other.
for example:
I want to add column with sum unit price on storage.
1) add on table `u_yf_istorages_products` structure new column `unit_total_price`
2) create view -> add on file modules/IStorage/models/RelationListView.php

`public function getHeaders()
	{
		$headerFields = parent::getHeaders();
		$headerFields1 = parent::getHeaders();
		if ('Products' == $this->getRelationModel()->get('modulename') && 'getManyToMany' == $this->getRelationModel()->get('name')) {
			$qtyInStock = new Vtiger_Field_Model();
			$qtyInStock->setModule(Vtiger_Module_Model::getInstance('Products'));
			$qtyInStock->set('name', 'qtyproductinstock');
			$qtyInStock->set('column', $qtyInStock->getName());
			$qtyInStock->set('label', 'FL_QTY_IN_STOCK');
			$qtyInStock->set('fieldDataType', 'double');
			$qtyInStock->set('fromOutsideList', true);
			if (App\Privilege::isPermitted('IStorages', 'SetQtyProducts')) {
				$qtyInStock->set('isEditable', true);
			}
			
			**$qtyInStock1 = new Vtiger_Field_Model();
			$qtyInStock1->setModule(Vtiger_Module_Model::getInstance('Products'));
			$qtyInStock1->set('name', 'unittotalprice');
			$qtyInStock1->set('column', $qtyInStock1->getName());
			$qtyInStock1->set('label', 'Wartość produktu:');
			$qtyInStock1->set('fieldDataType', 'double');
			$qtyInStock1->set('fromOutsideList', true);
			if (App\Privilege::isPermitted('IStorages', 'SetQtyProducts')) {
				$qtyInStock1->set('isEditable', false);**
				
			**//first time create new table	
			$headerFields['tab1'] = $qtyInStock;
			//second time create new table
			$headerFields['tab2'] = $qtyInStock1;**
			}
		}
		//return value for any relates module
		return $headerFields;`

3) add new query to file from modules/IStorage/models/GetManyToMany.php which will get value on new create column in on table u_yf_istorages_products. -> 
		$relatedModuleName = $this->relationModel->getRelationModuleName();
		if ('Products' === $relatedModuleName) {
			$parentModuleName = $this->relationModel->getParentModuleModel()->getName();
			$referenceInfo = \Vtiger_Relation_Model::getReferenceTableInfo($relatedModuleName, $parentModuleName);
			$this->relationModel->getQueryGenerator()->setCustomColumn(['qtyproductinstock' => $referenceInfo['table'] . '.qtyinstock']);
`			**$this->relationModel->getQueryGenerator()->setCustomColumn(['unittotalprice' => $referenceInfo['table'] . '.unit_total_price']);**`
		}
		parent::getQuery();

4) query to sync table with modul whats add products on storage:
`	\App\Db::getInstance()->createCommand("UPDATE u_yf_istorages_products 
	INNER JOIN u_yf_igrn_inventory 
	ON u_yf_istorages_products.relcrmid = u_yf_igrn_inventory.name 
	SET u_yf_istorages_products.unit_total_price = u_yf_igrn_inventory.price * u_yf_istorages_products.qtyinstock")->execute();`
I add on custom function with query on workflow in IGRN module.

**Comment by apcloic** [comment]
@crawe1 
Hi, thanks for sharing but I'm not sure that I understand correctly what you are trying to do.
Anyway, I think you may achieve the same thing using the GUI.
For example :
- [Create new field](https://gitdeveloper.yetiforce.com/index.php?module=LayoutEditor&parent=Settings&view=Index) in any module
- [Add columns](https://gitdeveloper.yetiforce.com/index.php?module=LayoutEditor&parent=Settings&view=Index&mode=showRelatedListLayout) to related list view for almost any module
- [Create Update field task](https://gitdeveloper.yetiforce.com/index.php?module=Workflows&parent=Settings&view=List) using Workflow module.
Also, you should avoid modifying files like modules/IStorage/models/GetManyToMany.php which may be overwritten by the next YF system update. It's better to [overwrite using loader mechanism](https://yetiforce.com/en/knowledge-base/documentation/developer-documentation/item/loader-and-autoloader-of-files).
Note that this feature needs to be enabled in the perfomance.php config file.
```
// Parameter that allows to disable file overwriting. After 
	// enabling it the system will additionally check whether the file exists in the custom directory.
	// Ex. custom/modules/Assets/Assets.php 
	'LOAD_CUSTOM_FILES' => true,
	'LOAD_CUSTOM_LANGUAGE' => true,
```
Regards

**Comment by crawe1** [comment]
@apcloic not in this example. One product many storage many quantity many price. In 99% u have right but not this time. :) I know, but i have change log. 
Maybe u help me about my last quetion. How do storage by FIFO model on yeti ?

**Reply by apcloic** [reply]
@crawe1 Ok, so I did not understand correctly ;)
Sorry I can't help for FIFO model as I don't use YF storage feature.
Regards
